protocols/imap/src/main/java/org/apache/james/imap/decode/parser/SetACLCommandParser.java

```diff
@@ -48,27 +48,41 @@ public SetACLCommandParser(StatusResponseFactory statusResponseFactory) {
 
     @Override
     protected ImapMessage decode(ImapRequestLineReader request, Tag tag, ImapSession session) throws DecodingException {
-        try {
-            final SetACLRequest.MailboxName mailboxName = new SetACLRequest.MailboxName(request.mailbox());
-            final Username identifier = Username.of(request.astring());
-            final String editModeAndRights = request.astring();
-            request.eol();
+        SetACLRequest.MailboxName mailboxName = new SetACLRequest.MailboxName(request.mailbox());
+        MailboxACL.EntryKey entryKey = MailboxACL.EntryKey.deserialize(request.astring());
+        String editModeAndRights = request.astring();
+        request.eol();
+
+        MailboxACL.ACLCommand aclCommand = MailboxACL.command()
+                .key(entryKey)
+                .mode(parseEditMode(editModeAndRights))
+                .rights(parseRights(editModeAndRights))
+                .build();
+
+        return new SetACLRequest(tag, mailboxName, aclCommand);
+    }
 
-            MailboxACL.EditMode editMode = MailboxACL.EditMode.REPLACE;
-            if (StringUtils.isNotEmpty(editModeAndRights)) {
-                switch (editModeAndRights.charAt(0)) {
-                    case MailboxACL.ADD_RIGHTS_MARKER:
-                        editMode = MailboxACL.EditMode.ADD;
-                        break;
-                    case MailboxACL.REMOVE_RIGHTS_MARKER:
-                        editMode = MailboxACL.EditMode.REMOVE;
-                        break;
-                }
-            }
+    private MailboxACL.EditMode parseEditMode(String editModeAndRights) {
+        if (StringUtils.isEmpty(editModeAndRights)) {
+            return MailboxACL.EditMode.REPLACE;
+        }
 
-            MailboxACL.Rfc4314Rights rights = MailboxACL.Rfc4314Rights.deserialize(editModeAndRights.substring(1));
+        return switch (editModeAndRights.charAt(0)) {
+            case MailboxACL.ADD_RIGHTS_MARKER -> MailboxACL.EditMode.ADD;
+            case MailboxACL.REMOVE_RIGHTS_MARKER -> MailboxACL.EditMode.REMOVE;
+            default -> MailboxACL.EditMode.REPLACE;
+        };
+    }
 
-            return new SetACLRequest(tag, mailboxName, identifier, editMode, rights);
+    private MailboxACL.Rfc4314Rights parseRights(String editModeAndRights) throws DecodingException {
+        if (StringUtils.isEmpty(editModeAndRights)) {
+            throw new DecodingException(HumanReadableText.INVALID_COMMAND, "SETACL command must include rights. If you want to remove rights for that user, please use a DELETEACL command.");
+        }
+        if (editModeAndRights.charAt(0) == MailboxACL.ADD_RIGHTS_MARKER || editModeAndRights.charAt(0) == MailboxACL.REMOVE_RIGHTS_MARKER) {
```

Comment:
Use Guava `CharMatcher` utility?

This will make the condition much more readable.